## Welcome to DEEP pipeline.

![DEEP Logo](./images/DEEP.png)

## README
This repository contains the scripts to run the DEEP pipeline which includes:

* Genome-wide scans for DNA methylation quantitative trait loci
* GWAS of DNAm derived phenotypes
* EWAS of genetic phenotypes

Please note the scripts in this repository are copyright of the Genetics of DNA Methylation Consortium and released with a [GPL v3 license](LICENSE).

### What is DEEP?

The [Diverse Epigenetic Epidemiology Partnership (DEEP)](https://www.deep-epigenetics.org/) is a partnership of 20 research groups from around the world that aims to improve global health by exploring the effects of genomic and environmental diversity on disease risk across the global population.

DEEP study researchers are interested in studying the differences in disease mechanism and disease risk prediction that are associated with DNA methylation. DNA methylation refers to modifications to DNA which help in turning genes on or off (transcription).The DEEP study ultimately aims to identify disease mechanisms and drug targets that work for all.

To give you an idea of the kind of analyses we are interested please review our previous publication: https://pubmed.ncbi.nlm.nih.gov/34493871/.

If you are interested in joining or finding out more please contact either Josine Min (Josine.Min@bristol.ac.uk) or Eilis Hannon (E.J.Hannon@exeter.ac.uk)

## Guidance for how to use this repository

Detailed instructions for running each stage of the pipeline [can be found here](https://github.com/genetics-of-dna-methylation-consortium/godmc_phase2/wiki). 
to change it

This includes details on how what software you need to have installed, and the required input data and their format. 

Whilst we have done a lot of testing of these scripts across a range of different datasets, we do anticipate that there will be bugs and edge cases we are yet to detect. All communication regarding this pipeline is to occur through this GitHub repository. If you email any of the developers they will direct you back to this repository. 

There are two ways to communicate with us. 

1. [Issues](https://github.com/MRCIEU/DEEP_mqtls/issues). This is the mechanism to report bugs or errors with the pipeline or documentation. Essentially anything which requires us to make edits to the content in the repository. There are more guidelines on how to submit issues below.
2. [Discussions](https://github.com/MRCIEU/DEEP_mqtls/discussions). This is the place to ask questions about the pipeline and analyses. 

## Submitting an issue

Thank you for helping improve our project! Before you submit an issue, please take a moment to review these guidelines to ensure your issue is addressed effectively.

### General Guidelines
**Search Existing Issues:** Before submitting a new issue, please search the existing issues (including closed issues) to see if your problem has already been reported. This helps avoid duplicates and allows us to focus on a single thread of discussion for each issue.

**Use Clear and Descriptive Titles:** Provide a concise summary of the issue in the title. This helps others understand and search for similar issues.

**Provide Detailed Information:**

**For Bugs**: Specify which script you were running when the error occured, the expected result, and the actual result. Provide any relevant logs or screenshots.

**For Documentation:** Specify the page & section where the documentation was misleading or confusing. 

## Guidance for contributing

The scripts in this repository were developed by the core DEEP team and individuals who had [approved external projects](http://www.godmc.org.uk/projects.html). A list can be found under the Acknowledgements section below. At present just this group are able to contribute to the pipeline - if you are interested in joining this group please contact either Josine Min (Josine.Min@bristol.ac.uk) or Eilis Hannon (E.J.Hannon@exeter.ac.uk).


## Acknowledgements

The developer group of these scripts are listed in alphabetical order below:

Carlos Ruiz Arenas  
Eilis Hannon  
Gibran Hemani  
Haotian Tang  
Josine Min  
Maria Soler Artigas  
Natalia Llonga  
Olalekan Awoniran  
Olivia Castellini  
Siyi Wang  
Sonja Rajic  
Thomas Jonkman  
Xiaopu Zhang  
