## Welcome to the DEEP pipeline.

![DEEP Logo](./images/DEEP.png)

## README
This repository contains the scripts to run the DEEP pipeline which includes:

* Genome-wide scans for DNA methylation quantitative trait loci
* GWAS of DNAm derived phenotypes
* EWAS of genetic phenotypes
* Epigenetic prediction of human phenotypes

Please note the scripts in this repository are copyright of the Diverse Epigenetic Epidemiology Partnership and released with a [GPL v3 license](LICENSE).

### What is DEEP?

The [Diverse Epigenetic Epidemiology Partnership (DEEP)](https://www.deep-epigenetics.org/) is a partnership of research groups located globally. DEEP aims to improve global health by exploring the effects of genomic and environmental diversity on disease risk across the global population.

DEEP study researchers are interested in studying the differences in disease mechanism and disease risk prediction that are associated with DNA methylation. DNA methylation refers to modifications to DNA which help in turning genes on or off (transcription).The DEEP study ultimately aims to identify disease mechanisms and drug targets that work for all.

If you are interested in joining or finding out more, please contact either Josine Min (Josine.Min@bristol.ac.uk) or Hannah Elliott (Hannah.Elliott@Bristol.ac.uk)

## Guidance for how to use this repository

Detailed instructions for running each stage of the pipeline [can be found here](https://github.com/MRCIEU/DEEP_mqtls/wiki). 
to change it

This includes details on how what software you need to have installed, and the required input data and their format. 

Whilst we have done a lot of testing of these scripts across a range of different datasets, there may still be bugs and edge cases. 

When running this pipeline, there are two ways to share errors or comments: 

1. [Issues](https://github.com/MRCIEU/DEEP_mqtls/issues). This is the mechanism to report bugs or errors with the pipeline or documentation. Essentially anything which requires us to make edits to the content in the repository. There are more guidelines on how to submit issues below.
2. [Discussions](https://github.com/MRCIEU/DEEP_mqtls/discussions). This is the place to ask questions or make suggestions about the pipeline and analyses. 

## Submitting an issue

Before you submit an issue, please take a moment to review these guidelines to ensure your issue is addressed effectively.

### General Guidelines
**Search Existing Issues:** Before submitting a new issue, please search the existing issues (including closed issues) to see if your problem has already been reported. This helps avoid duplicates and allows us to focus on a single thread of discussion for each issue.

**Use Clear and Descriptive Titles:** Provide a concise summary of the issue in the title. This helps others understand and search for similar issues.

**Provide Detailed Information:**

**For Bugs**: Specify which script you were running when the error occured, the expected result, and the actual result. Provide any relevant logs or screenshots.

**For Documentation:** Specify the page & section where the documentation was misleading or confusing. 

## Guidance for contributing

The scripts in this repository were developed by the core DEEP team, building on previous work from GoDMC. A list can be found under the Acknowledgements section below.


## Acknowledgements

The developer group of these scripts are listed in alphabetical order below:

Hannah Elliott
Josine Min 
Haotian Tang
Sarah Watkins
  
Scripts were cloned from the GoDMC2 pipeline (https://github.com/genetics-of-dna-methylation-consortium/godmc_phase2) and adopted for DEEP purposes. 
